python/tank/pipelineconfig_factory.py

```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
+         ]
+    )
+
+    # Introspect the TANK_CURRENT_PC env var to determine which pipeline configuration
+    # the current sgtk import belongs to.
+    #
+    # if the call returns a path, we are running a localized pipeline configuration, e.g.
+    # the pipeline configuration has its own associated core. If it returns None,
+    # we are running a shared core, e.g. a core API which is used by multiple
+    # pipeline configurations.
     config_context_path = _get_configuration_context()
 
     if config_context_path:
 
-        # we are running the tank command or python core API directly from a configuration
-        #
-        # now if this tank command is associated with the path, the registered path should be in
-        # in the pipeline configuration data coming from
-        if config_context_path not in [x["path"] for x in all_pc_data]:
-
-            pcs_msg = ", ".join([
-                            "'%s' (Pipeline config id %s, Project id %s)" % (x["path"], x["id"], x["project_id"])
-                            for x in all_pc_data])
-
-            raise TankError("You are trying to start Toolkit using the pipeline configuration "
-                            "located in '%s'. The path '%s' you are trying to load is not "
-                            "associated with that configuration. Instead, it is "
-                            "associated with the following pipeline configurations: %s. "
-                            "Please use the tank command or Toolkit API in any of those "
-                            "locations in order to continue. This error can occur if you "
-                            "have moved a configuration manually rather than using "
-                            "the 'tank move_configuration command'. It can also occur if you "
-                            "are trying to use a tank command associated with one Project "
-                            "to try to operate on a Shot or Asset that belongs to another "
-                            "project." % (config_context_path, path, pcs_msg))
-
-        # okay so this pipeline config is valid!
-        return PipelineConfiguration(config_context_path)
+        # --- THE LOCALIZED CORE CASE ----
+
+        # This is the localized case where the imported code has a 1:1 correspondence
+        # with the pipeline configuration. Now we need to verify that the path is compatible
+        # with this configuration, or raise an exception.
+
+        # create an instance to represent our path
+        pipeline_configuration = PipelineConfiguration(config_context_path)
+
+        # get pipeline config shotgun id
+        pc_id = pipeline_configuration.get_shotgun_id()
+
+        # find the pipeline config in our list of configs. If we cannot find it, we
+        # don't have a match beetween the code that is being run and the config
+        # we are trying to start up.
+
+        if pc_id not in [x["id"] for x in all_pc_data]:
```

Comment:
this is the actual fix - previously we were comparing by path, but i realised that we could compare by id and all the problems would go away.